File: cairnkit/notify.py
"""Notification dispatch (M10) — reach people at key moments.

v1 channel is a Feishu/Lark webhook bot, triggered by hooks at CLARIFY-pending / blocked /
DONE. The channel is pluggable and the webhook URL is read from an environment variable named
in config — never hardcoded. With no webhook configured (or a send failure) it degrades to a
local file reminder so the workflow never breaks.
"""

from __future__ import annotations

import json
import os
import urllib.request

EVENTS = ("clarify_pending", "blocked", "done", "archived", "arch_review")
_LOCAL_FILE = ".cairnkit/notifications.log"


def build_message(event: str, project: str, detail: str = "") -> str:
    return f"[cairnkit:{project}] {event}" + (f" — {detail}" if detail else "")


def notify(event: str, config, detail: str = "", channel: str = "feishu") -> dict:
    """Send a notification, degrading to a local file when no webhook is configured/reachable."""
    text = build_message(event, config.project, detail)
    env_name = getattr(config, "notify_webhook_env", None)
    url = os.environ.get(env_name) if env_name else None
    if not url:
        return _local(config, text, reason="no webhook configured")
    try:
        _send_feishu(url, text)
        return {"sent": True, "channel": channel}
    except Exception as exc:  # network/HTTP failure -> degrade, never break the flow
        return _local(config, text, reason=f"send failed: {exc}")


def _send_feishu(url: str, text: str) -> None:
    payload = json.dumps({"msg_type": "text", "content": {"text": text}}).encode("utf-8")
    req = urllib.request.Request(  # noqa: S310 - url comes from operator config/env
        url, data=payload, headers={"Content-Type": "application/json"}
    )
    urllib.request.urlopen(req, timeout=10).read()  # noqa: S310


def _local(config, text: str, reason: str) -> dict:
    path = config.root / _LOCAL_FILE
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as fh:
        fh.write(text + "\n")
    return {"sent": False, "local": str(path.relative_to(config.root)), "reason": reason}

File: src/notify.rs
//! Notification dispatch (M10). Webhook via `curl` (no HTTP dependency); degrade to a local file.

use crate::config::Config;
use serde_json::{json, Value};
use std::process::Command;

const LOCAL_FILE: &str = ".cairnkit/notifications.log";

pub fn build_message(event: &str, project: &str, detail: &str) -> String {
    if detail.is_empty() {
        format!("[cairnkit:{project}] {event}")
    } else {
        format!("[cairnkit:{project}] {event} — {detail}")
    }
}

pub fn notify(event: &str, config: &Config, detail: &str, channel: &str) -> Value {
    let text = build_message(event, &config.project, detail);
    let url = config
        .notify_webhook_env
        .as_ref()
        .and_then(|name| std::env::var(name).ok());
    match url {
        None => local(config, &text, "no webhook configured"),
        Some(u) => match send_feishu(&u, &text) {
            Ok(()) => json!({"sent": true, "channel": channel}),
            Err(e) => local(config, &text, &format!("send failed: {e}")),
        },
    }
}

fn send_feishu(url: &str, text: &str) -> Result<(), String> {
    let payload = json!({"msg_type": "text", "content": {"text": text}}).to_string();
    let out = Command::new("curl")
        .args([
            "-sS",
            "-X",
            "POST",
            "-H",
            "Content-Type: application/json",
            "-d",
            &payload,
            url,
        ])
        .output()
        .map_err(|e| e.to_string())?;
    if out.status.success() {
        Ok(())
    } else {
        Err(String::from_utf8_lossy(&out.stderr).trim().to_string())
    }
}

fn local(config: &Config, text: &str, reason: &str) -> Value {
    use std::io::Write;
    let path = config.root.join(LOCAL_FILE);
    if let Some(parent) = path.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    if let Ok(mut f) = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&path)
    {
        let _ = writeln!(f, "{text}");
    }
    let rel = path
        .strip_prefix(&config.root)
        .unwrap_or(&path)
        .to_string_lossy()
        .to_string();
    json!({"sent": false, "local": rel, "reason": reason})
}
